### PR TITLE
HPCC-22316 Extend WU Iterators to provide errors, warnings and analysis

### PR DESCRIPTION
**Pull request by shamser**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [x] Test my draft Pull Request.

### CODE REVIEW

**Comment by AttilaVamos** [comment]
@shamser ECLAgent cored.
```
Core was generated by `eclagent WUID=W20190614-124243-4 DALISERVERS=10.241.40.11:7070'.
Program terminated with signal 11, Segmentation fault.
#0  0x00007f697cb4694d in NotesIterator (_filter=..., wu=0x10e61b0, this=0x1128670) at /mnt/disk1/home/vamosax/smoketest/PR-12670/HPCC-Platform/common/workunit/workunit.cpp:1045
1045	            if (strlen(exception.queryScope())!=0)
 
Thread 1 (Thread 0x7f697df00140 (LWP 13761)):
#0  0x00007f697cb4694d in NotesIterator (_filter=..., wu=0x10e61b0, this=0x1128670) at /mnt/disk1/home/vamosax/smoketest/PR-12670/HPCC-Platform/common/workunit/workunit.cpp:1045
#1  CLocalWorkUnit::getScopeIterator (this=0x10e61b0, filter=...) at /mnt/disk1/home/vamosax/smoketest/PR-12670/HPCC-Platform/common/workunit/workunit.cpp:8781
#2  0x00007f697d77893e in WorkunitAnalyser::collateWorkunitStats (this=this@entry=0x7ffc6416d030, workunit=workunit@entry=0x10f7020, filter=...) at /mnt/disk1/home/vamosax/smoketest/PR-12670/HPCC-Platform/common/wuanalysis/anawu.cpp:372
#3  0x00007f697d779054 in WorkunitAnalyser::analyse (this=this@entry=0x7ffc6416d030, wu=wu@entry=0x10f7020) at /mnt/disk1/home/vamosax/smoketest/PR-12670/HPCC-Platform/common/wuanalysis/anawu.cpp:350
#4  0x00007f697d7791f4 in analyseWorkunit (wu=0x10f7020, options=...) at /mnt/disk1/home/vamosax/smoketest/PR-12670/HPCC-Platform/common/wuanalysis/anawu.cpp:413
#5  0x00007f697da9d852 in EclAgent::doProcess (this=this@entry=0x7ffc6416d330) at /mnt/disk1/home/vamosax/smoketest/PR-12670/HPCC-Platform/ecl/eclagent/eclagent.cpp:1959
#6  0x00007f697da9eb90 in eclagent_main (argc=argc@entry=3, argv=argv@entry=0x7ffc6416d6d8, wuXML=wuXML@entry=0x0, standAloneExe=standAloneExe@entry=false) at /mnt/disk1/home/vamosax/smoketest/PR-12670/HPCC-Platform/ecl/eclagent/eclagent.cpp:3595
#7  0x0000000000401bd0 in main (argc=3, argv=0x7ffc6416d6d8) at /mnt/disk1/home/vamosax/smoketest/PR-12670/HPCC-Platform/ecl/eclagent/eclagentmain.cpp:79
 ```
Core and trace files are uploaded to JIRA.

File: common/workunit/workunit.cpp [hunk 1]
```diff
@@ -1023,6 +1023,151 @@ class CConstGraphProgressScopeIterator : public CInterfaceOf<IConstWUScopeIterat
...
+class NotesIterator : public CInterfaceOf<IConstWUScopeIterator>
+{
+public:
+    NotesIterator(const IConstWorkUnit * wu, const ScopeFilter & _filter)
+    {
+        Owned<IConstWUExceptionIterator> exceptions = &wu->getExceptions();
+        ForEach(*exceptions)
+        {
+            IConstWUException & exception = exceptions->query();
+            if (strlen(exception.queryScope())!=0)
```
Comment (ghalliday) [inline common/workunit/workunit.cpp]: Can queryScope be null?  I think we said that blank scopes would be allowed, but null scopes would be ignored.

File: common/workunit/workunit.cpp [hunk 2]
```diff
@@ -8626,6 +8776,11 @@ IConstWUScopeIterator & CLocalWorkUnit::getScopeIterator(const WuScopeFilter & f
             compoundIter->addIter(workflowIter);
         }
     }
+    //if (sources & SSFsearchWorkflow)
```
Comment (ghalliday) [inline common/workunit/workunit.cpp]: Should check sources & SSFsearchExceptions.  But will need to check that flag is correctly set up from the filter.  E.g. in finishedFilter, inside the if (!sourceFlags):

if (!(properties & (PTnotes)))
         sourceFlags &= ~(SSFsearchExceptions);

**Review by ghalliday — COMMENTED** [review]
@shamser a few minor comments, but everything looks good - structure is what I was expecting.

@GordonSmith please can we have feedback on the interface - so we have a chance of merging the change into 7.4.

File: esp/scm/ws_workunits.ecm [hunk 3]
```diff
@@ -2204,12 +2205,25 @@ ESPstruct [nil_remove] WUResponseProperty
     [optional] string CreatorType;  // What type of engine created it.
 };
 
+ESPstruct [nil_remove] WUResponseNote
+{
+    [optional] string Source;       // Thor, Eclcc, Analyser etc
+    [optional] string Message;
+    [optional] int64 ErrorCode;
```
Comment (GordonSmith) [inline esp/scm/ws_workunits.ecm]: Is this really an int64?

File: esp/scm/ws_workunits.ecm [hunk 4]
```diff
@@ -2156,6 +2156,7 @@ ESPStruct WUPropertiesToReturn
     bool AllHints(false);
     bool AllScopes(false);
     bool AllProperties(false);
+    [min_ver("1.76")] bool AllNotes(false);
```
Comment (GordonSmith) [inline esp/scm/ws_workunits.ecm]: I assume this means either get them or not (no option to cherry pick some)?  Which I am pretty sure will be fine.

File: esp/scm/ws_workunits.ecm (hunk 3 above)
Comment (GordonSmith) [reply esp/scm/ws_workunits.ecm]: (JS Can't handle a full int64...)

**Review by GordonSmith — APPROVED** [review]
Two comments, but otherwise looks fine.

File: esp/scm/ws_workunits.ecm (hunk 3 above)
Comment (shamser) [reply esp/scm/ws_workunits.ecm]: The error code is 32bit unsigned - I'll change the type.

File: common/workunit/workunit.cpp (hunk 1 above)
Comment (shamser) [reply common/workunit/workunit.cpp]: Apparently, yes.  Fixing.

File: esp/scm/ws_workunits.ecm (hunk 4 above)
Comment (shamser) [reply esp/scm/ws_workunits.ecm]: Yes, at the moment, it is all or nothing as far as the notes.  We could add filtering later if required.

Comment (shamser) [reply esp/scm/ws_workunits.ecm]: Yes - at least for now.  If filtering would be useful, we can add later.

**Comment by shamser** [comment]
@ghalliday Please can you review.

**Comment by ghalliday** [comment]
Need to add { PTnotes, "notes" } to propertyMappings in line 2615

File: tools/wutool/wutool.cpp [hunk 5]
```diff
@@ -174,6 +174,27 @@ class ScopeDumper : public IWuScopeVisitor
...
+        StringBuffer xml;
+        SCMStringBuffer source, message, filename, timestamp, scope;
+
+        exception.getExceptionSource(source);
+        exception.getExceptionMessage(message);
+        exception.getTimeStamp(timestamp);
+        exception.getExceptionFileName(filename);
+
+        // TODO: check if anything below needs encoding
+        xml.appendf("<note source='%s' message='%s' timestamp='%s' exceptionCode='%u' severity='%u' scope='%s' cost='%u'",
```
Comment (ghalliday) [inline tools/wutool/wutool.cpp]: trivial: no need to include the scope - it is already output.

File: common/workunit/workunit.cpp [hunk 6]
```diff
@@ -1023,6 +1023,151 @@ class CConstGraphProgressScopeIterator : public CInterfaceOf<IConstWUScopeIterat
...
+{
+    return compareScopeName(left.queryScope(), right.queryScope());
+}
+
+int compareNoteScopeOrder(IInterface * const * left, IInterface * const * right)
+{
+    return compareNoteScopeOrder(*static_cast<IConstWUException *>(*left), *static_cast<IConstWUException *>(*right));
+}
+
+class NotesIterator : public CInterfaceOf<IConstWUScopeIterator>
```
Comment (ghalliday) [inline common/workunit/workunit.cpp]: At some point some of this code could be commoned up with the global stats - it it was thought worthwhile.  Probably when the next iterator comes along.

File: system/jlib/jstatcodes.h [hunk 7]
```diff
@@ -69,6 +69,7 @@ enum StatisticScopeType
     SSTfunction,                        // a function call
     SSTworkflow,
     SSTchildgraph,
+    SSTnote,
```
Comment (ghalliday) [inline system/jlib/jstatcodes.h]: It isn't really a scope type - it is a type of property.  It doesn't exist independently of a scope.

File: common/workunit/workunit.cpp [hunk 8]
```diff
@@ -2917,6 +3066,11 @@ void WuScopeFilter::finishedFilter()
             if (!(properties & (PTattributes|PThints)))
                 sourceFlags &= ~(SSFsearchGraph|SSFsearchWorkflow);
         }
+
+        if (!(properties & PTnotes))
+            sourceFlags &= ~SSFsearchExceptions;
+        else
```
Comment (ghalliday) [inline common/workunit/workunit.cpp]: I don't think the else is needed.  When I tested with properties[note] (after adding it to the table) the value was output by wutool.

Comment (ghalliday) [reply common/workunit/workunit.cpp]: Is the problem that w1, w1:graph1 and w1:graph1:sg1 are not output?

File: common/workunit/workunit.cpp [hunk 9]
```diff
@@ -3113,6 +3271,8 @@ StringBuffer & WuScopeFilter::describe(StringBuffer & out) const
                 props.append(",hint");
             if (properties & PTscope)
                 props.append(",scope");
+            if (properties & PTnotes)
+                props.append(",notes");
```
Comment (ghalliday) [inline common/workunit/workunit.cpp]: should be ,note for consistency of singular/plural

**Review by ghalliday — COMMENTED** [review]
@shamser a few fairly minor comments.  Looks close to being ready to merge.

File: common/workunit/workunit.cpp (hunk 6 above)
Comment (shamser) [reply common/workunit/workunit.cpp]: Noted

File: common/workunit/workunit.cpp (hunk 8 above)
Comment (shamser) [reply common/workunit/workunit.cpp]: Nope, the problem is that if only the 'notes' is selected, then all the other iterators are added as well. Also, "global" and "workflow" scopes (that have no notes) are being returned when only "notes" are requested

**Comment by shamser** [comment]
@ghalliday

File: common/workunit/workunit.cpp [hunk 10]
```diff
@@ -1023,6 +1023,153 @@ class CConstGraphProgressScopeIterator : public CInterfaceOf<IConstWUScopeIterat
...
+    {
+        return nullptr;
+    }
+
+    virtual const char * queryHint(const char * kind) const override
+    {
+        return nullptr;
+    }
+
+    IMPLEMENT_IINTERFACE_USING(CInterfaceOf<IConstWUScopeIterator>);
```
Comment (ghalliday) [inline common/workunit/workunit.cpp]: minor: This line can be deleted

File: common/workunit/workunit.cpp (hunk 8 above)
Comment (ghalliday) [reply common/workunit/workunit.cpp]: How are you requesting the information?

File: common/workunit/workunit.cpp [hunk 11]
```diff
@@ -2956,7 +3112,7 @@ void WuScopeFilter::finishedFilter()
     if (include.nestedDepth == 0)
     {
         if (scopeFilter.compareDepth(3) > 0)    // 3 is larger than any scope in the filter
-            sourceFlags &= ~(SSFsearchGraphStats);
+            sourceFlags &= ~(SSFsearchGraphStats|SSFsearchExceptions);
```
Comment (ghalliday) [inline common/workunit/workunit.cpp]: Why is this true?  I would expect to always need to search the exceptions.

File: system/jlib/jexcept.cpp [hunk 12]
```diff
@@ -56,6 +56,16 @@
 #define LINUX_SIGNAL_EXCEPTION
 #endif
 
+const char * errorSeverityString[] = {"Information", "Warning", "Error", "Alert", "Ignore", "Fatal", "Unknown"};
+
+const char * querySeverityString(ErrorSeverity errorSeverity)
+{
+    if (errorSeverity<sizeof(errorSeverityString))
```
Comment (ghalliday) [inline system/jlib/jexcept.cpp]: sizeof is the wrong test.  You can use _elements_in(x) (from jlib.hpp)

**Review by ghalliday — COMMENTED** [review]
@shamser looks close.  A few remaining comments.

File: common/workunit/workunit.cpp (hunk 11 above)
Comment (shamser) [reply common/workunit/workunit.cpp]: Originally, I thought all issues would be related to activities.  This is not true, so this change is needed.

File: common/workunit/workunit.cpp (hunk 8 above)
Comment (shamser) [reply common/workunit/workunit.cpp]: I had PropertiesToReturn=PTnotes,  maxDepth=0 and nestedDepth=9999.  My mistake I should have set maxDepth=999. My mistake.  (SSFsearchGlobalStats not required when searching PTnotes specified)

File: common/workunit/workunit.cpp [hunk 13]
```diff
@@ -2267,6 +2265,7 @@ class CompoundStatisticsScopeIterator : public CInterfaceOf<IConstWUScopeIterato
                 if (nesting <= filter.include.nestedDepth)
                     include = true;
             }
+
```
Comment (shamser) [inline common/workunit/workunit.cpp]: Restoring line removed by an earlier commit.

**Pull request merged**